text/0076-rawkv-batch-export.md

```diff
@@ -0,0 +1,104 @@
+- data export, which we want to provide in this RFC.
+
+`RawKV Batch Export` should meet the following requirements:
+
+1. the export process should be as fast as possible,
+2. the impact to the online serving (P99 latency) should be under control,
+3. the export speed can be improved in a linear way by adding more tikv instances,
+4. the export framework should be able to process Big Data (500GB ~ 5TB) in a short period (from minutes to several hours).
+
+## Detailed design
+
+The solution proposed in this RFC is as follows:
+
+1. Using [RawKV backup](https://docs.pingcap.com/zh/tidb/stable/use-br-command-line-tool#raw-kv-%E5%A4%87%E4%BB%BD%E5%AE%9E%E9%AA%8C%E6%80%A7%E5%8A%9F%E8%83%BD) provide by BR, TiKV can backup SST files to S3. But HDFS is not supported yet, which is describe in [Support upload SST to HDFS](#support-upload-sst-to-hdfs).
+2. After uploading SST files to S3 / HDFS, we will use Spark to decode the SST files to Key-Value pairs with the help of [the RocksDB SstFileReader API](https://github.com/facebook/rocksdb/blob/6.23.fb/java/src/main/java/org/rocksdb/SstFileReader.java). And then the Key-Value pairs will be convert to user firendly file formats like parquet, which can be controlled by user code. After that the parquet file will finally upload to S3 / HDFS.
+
+![Batch Export](../media/batch-export.png)
+
+### Support upload SST to HDFS
+
+Most of the non-cloud users choose to use HDFS as their big data infra. In order to support these users to choose RawKV in their feature store, HDFS must be supported.
+
+As HDFS Client is implemented using java, a problem occures: "how to call java in rust?". There are some options:
+
+- [HDFS Shell Command](https://hadoop.apache.org/docs/current/hadoop-project-dist/hadoop-common/FileSystemShell.html#put)
+- [C API libhdfs](https://hadoop.apache.org/docs/stable/hadoop-project-dist/hadoop-hdfs/LibHdfs.html)
+- [j4rs](https://github.com/astonbitecode/j4rs)
+- [webhdfs rest api](https://hadoop.apache.org/docs/r1.0.4/webhdfs.html)
+
+`HDFS Shell Command` is chosen, considering
+
+- Deploy HDFS client environment on TiKV machines is easy, cause all the users have their own deploy tools.
+- Beside apache HDFS, HDFS compatible systems like [AWS S3](https://hadoop.apache.org/docs/current/hadoop-aws/tools/hadoop-aws/index.html)、[OSS](https://developer.aliyun.com/article/292792)、[Huawei OBSA](https://github.com/huaweicloud/obsa-hdfs) should also be considered. `HDFS Shell Command` is the most compatible solution.
+- Calling HDFS Shell Command is the simplest solution.
+
+#### Config
+
+We will add some new configration items:
+
+1. `HADOOP_HOME`: let TiKV know how to find the hdfs shell command.
+2. `HADOOP_LINUX_USER`: TiKV will run the hdfs shell command under this linux user. TiKV will use the current linux user if not provided.
+
+The configration can be passed by
+
+- system environment
+- TiKV configuration file
```

Comment:
What key should be used when using configuration file?

Reply:
- system environment: `HADOOP_HOME`、`HADOOP_LINUX_USER`
- TiKV configuration file: `backup.hadoop.home`、`backup.hadoop.linux-user`